kani-compiler/src/kani_middle/resolve.rs

```diff
@@ -6,426 +6,363 @@
-                }
-            })
+pub fn resolve_fn(
+    tcx: TyCtxt,
+    current_module: LocalDefId,
+    path_str: &str,
+) -> Result<DefId, ResolveError> {
+    let result = resolve_path(tcx, current_module, path_str);
+    match result {
+        Ok(def_id) => {
+            let def_kind = tcx.def_kind(def_id);
+            if matches!(def_kind, DefKind::AssocFn | DefKind::Fn) {
+                Ok(def_id)
+            } else {
+                let description = format!(
+                    "expected function / method, found {} `{}`",
+                    def_kind.descr(def_id),
+                    tcx.def_path_str(def_id)
+                );
+                Err(ResolveError { msg: description })
+            }
         }
+        err => err,
     }
 }
 
-/// The segments of a path.
-type Segments = VecDeque<String>;
+/// Attempts to resolve a simple path (in the form of a string) to a `DefId`.
+/// The current module is provided as an argument in order to resolve relative
+/// paths.
+///
+/// Note: This function was written to be generic, however, it has only been tested for functions.
+fn resolve_path(
+    tcx: TyCtxt,
+    current_module: LocalDefId,
+    path_str: &str,
+) -> Result<DefId, ResolveError> {
+    let _span = tracing::span!(tracing::Level::DEBUG, "path_resolution").entered();
 
-/// Generates the string representation of the path composed of these segments
-/// (it is inefficient, but we use it only in debugging output).
-fn segments_to_string(segments: &Segments) -> String {
-    segments.iter().cloned().collect::<Vec<_>>().join("::")
+    let path = resolve_prefix(tcx, current_module, path_str)?;
+    path.segments.into_iter().try_fold(path.base, |base, name| {
+        debug!(?base, ?name, "resolve_path");
+        let def_kind = tcx.def_kind(base);
+        let next_item = match def_kind {
+            DefKind::ForeignMod | DefKind::Mod => resolve_in_module(tcx, base, &name),
+            DefKind::Struct | DefKind::Enum | DefKind::Union => resolve_in_type(tcx, base, &name),
+            kind => {
+                debug!(?base, ?kind, "resolve_path: unexpected item");
+                Err(ResolveError {
+                    msg: format!(
+                        "expected module, found {} `{}`",
+                        def_kind.descr(base),
+                        tcx.def_path_str(base)
+                    ),
+                })
+            }
+        };
+        next_item
+    })
 }
 
-/// The "starting point" for a path.
-#[derive(Debug)]
-enum Base {
-    /// Indicates an external path.
-    ExternPrelude,
-    /// Indicates a path that may be local (and must be local if
-    /// `may_be_external_path` is false) and should be resolved relative to the
-    /// module identified by `id`.
-    LocalModule { id: LocalDefId, may_be_external_path: bool },
+/// Provide information about where the resolution failed.
+/// Todo: Add error message.
+pub struct ResolveError {
+    pub msg: String,
+}
```

Comment:
Can we make this an `enum` with variants for each kind of error you can get? That way, all formatting (which is now done each time you construct an error) can be put in the `fmt` function.

Reply:
yeah, we could... I was just trying to do a quick hack